src/devinterp/utils.py

```diff
@@ -1,31 +1,91 @@
 
 
-def plot_trace(trace, y_axis, x_axis='step', title=None, plot_mean=True, plot_std=True, fig_size=(12, 9), true_lc = None):
+def plot_trace(
+    trace,
+    y_axis,
+    x_axis="step",
+    title=None,
+    plot_mean=True,
+    plot_std=True,
+    fig_size=(12, 9),
+    true_lc=None,
+):
     num_chains, num_draws = trace.shape
     sgld_step = list(range(num_draws))
     if true_lc:
-        plt.axhline(y=true_lc, color='r', linestyle='dashed')
+        plt.axhline(y=true_lc, color="r", linestyle="dashed")
     # trace
     for i in range(num_chains):
         draws = trace[i]
-        plt.plot(sgld_step, draws, linewidth=1, label=f'chain {i}')
+        plt.plot(sgld_step, draws, linewidth=1, label=f"chain {i}")
 
     # mean
     mean = np.mean(trace, axis=0)
-    plt.plot(sgld_step, mean, color='black', linestyle='--', linewidth=2, label='mean', zorder=3)
-    
+    plt.plot(
+        sgld_step,
+        mean,
+        color="black",
+        linestyle="--",
+        linewidth=2,
+        label="mean",
+        zorder=3,
+    )
+
     # std
     std = np.std(trace, axis=0)
-    plt.fill_between(sgld_step, mean - std, mean + std, color='gray', alpha=0.3, zorder=2)
+    plt.fill_between(
+        sgld_step, mean - std, mean + std, color="gray", alpha=0.3, zorder=2
+    )
 
     if title is None:
-        title = f'{y_axis} values over sampling draws'
-    plt.legend(loc='upper left', bbox_to_anchor=(1, 1))    
+        title = f"{y_axis} values over sampling draws"
+    plt.legend(loc="upper left", bbox_to_anchor=(1, 1))
     plt.title(title)
     plt.xlabel(x_axis)
     plt.ylabel(y_axis)
     plt.figure(figsize=fig_size)
     plt.tight_layout()
     plt.show()
+
+
+def optimal_temperature(data):
+    if isinstance(data, DataLoader):
+        return len(data.dataset) / np.log(len(data.dataset))
+    elif isinstance(data, Dataset) or isinstance(data, list):
+        return len(data) / np.log(len(data))
+    elif isinstance(data, int):
+        return data / np.log(data)
+    else:
+        raise NotImplementedError(f'Temperature for data type {type(data)} not implemented, use Dataloader or Dataset instead.')
+
+
+def get_init_loss_one_batch(dataloader, model, criterion, device):
+    model = model.to(device)
+    model.train() # to make sure we're using train loss, comparable to train loss of sampler()
+    with torch.no_grad():
+        xs, ys = next(iter(dataloader))
+        xs, ys = xs.to(device), ys.to(device)
+        y_preds = model(xs)
+        loss = criterion(y_preds, ys).detach().item()
+    return loss
+
+
+def get_init_loss_full_batch(dataloader, model, criterion, device):
+    model = model.to(device)
+    model.train()
+    loss = 0.
+    with torch.no_grad():
+        for xs, ys in dataloader:
+            xs, ys = xs.to(device), ys.to(device)
+            y_preds = model(xs)
+            loss += criterion(y_preds, ys).detach().item()
+    return loss
+
+def transformers_cross_entropy(inputs, outputs):
```

Comment:
haven't looked at the notebooks yet, but if this is meant to be a utility loss function to be passed as the criterion for things like llc estimation or `get_init_loss` functions, then i've been using this implementation (for training as well): https://github.com/neelnanda-io/TransformerLens/blob/main/transformer_lens/utils.py#L121

in particular i think at minimum you need the token offset between the logits and the outputs

Reply:
nah i think it's fine to delete it, we can defer the decision about including architecture-specific utils for another day